<a href="https://colab.research.google.com/github/HongYeongPyo/first-repository/blob/master/5%EB%A8%B8%EC%8B%A0_3%EB%B6%84%EB%A5%98_2%EC%9D%98%EC%82%AC%EA%B2%B0%EC%A0%95%EB%82%98%EB%AC%B4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 3-2 의사결정나무
- 필요한 라이브러리 가져오기
- sklearn 라이브러리에 내장된 데이터 불러오기

In [14]:
import pandas as pd
from sklearn.model_selection import  train_test_split
from sklearn.metrics import accuracy_score

In [15]:
# 데이터 생성
from sklearn.datasets import load_breast_cancer

load_breast_cancer()

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [20]:

iris = load_breast_cancer()
iris


{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [4]:
def make_dataset():
    iris = load_breast_cancer()
    df = pd.DataFrame(iris.data, columns = iris.feature_names)
    df['target'] = iris.target

    X_train, X_test, y_train, y_test = train_test_split(
        df.drop('target', axis=1), df['target'], test_size=0.5, random_state=1004
    )

X_train, X_test, y_train, y_test = make_dataset()
X_train.shape, X_test.shape, y_train.shape, y_test.shape


TypeError: cannot unpack non-iterable NoneType object

## 다음 코딩은 위의 코딩에서 오류(Aiffel LMS) 수정이다.

### 오류 원인 :
- TypeError: cannot unpack non-iterable NoneType object는 make_dataset() 함수가 None을 반환하고 있기 때문에 발생한 것입니다.

- train_test_split 함수의 반환값을 변수에 할당하려고 하는데, None이 반환되면 이를 여러 변수에 나눠서 할당할 수 없습니다.


### 오류 수정 :
- make_dataset() 함수가 데이터를 제대로 반환하도록 수정해야 합니다.
- train_test_split 함수에서 데이터를 반환하는 부분을 제대로 작성해야 합니다.


### make_dataset() 함수 호출 후 반환값 확인(None 값 확인)
- dataset = make_dataset()
- print("Returned Value: ", dataset)
- print("Type of Returned Value: ", type(dataset))

In [10]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
import pandas as pd

def make_dataset():
    iris = load_breast_cancer()
    df = pd.DataFrame(iris.data, columns=iris.feature_names)
    df['target'] = iris.target

    # train_test_split 함수의 결과를 반환하도록 수정
    X_train, X_test, y_train, y_test = train_test_split(
        df.drop('target', axis=1), df['target'], test_size=0.5, random_state=1004
    )
    return X_train, X_test, y_train, y_test

X_train, X_test, y_train, y_test = make_dataset()
X_train.shape, X_test.shape, y_train.shape, y_test.shape


((284, 30), (285, 30), (284,), (285,))

In [12]:
# make_dataset() 함수 호출 후 반환값 확인
dataset = make_dataset()
print("Returned Value: ", dataset)
print("Type of Returned Value: ", type(dataset))

Returned Value:  (     mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
544       13.870         20.70           89.77      584.8          0.09578   
508       16.300         15.70          104.70      819.8          0.09427   
365       20.440         21.78          133.80     1293.0          0.09150   
176        9.904         18.06           64.60      302.4          0.09699   
382       12.050         22.72           78.75      447.8          0.06935   
..           ...           ...             ...        ...              ...   
239       17.460         39.28          113.40      920.6          0.09812   
325       12.670         17.30           81.25      489.9          0.10280   
31        11.840         18.70           77.93      440.6          0.11090   
523       13.710         18.68           88.73      571.0          0.09916   
2         19.690         21.25          130.00     1203.0          0.10960   

     mean compactness  mean concavity  mean c

### 위의 명령 실형 결과 설명:
- make_dataset() 함수의 실행
- make_dataset() 함수는 load_breast_cancer()로부터 데이터를 불러오고,
- 이를 Pandas DataFrame으로 변환 후 train_test_split으로 데이터를 훈련 세트와 테스트 세트로 나누는 함수입니다.
- 함수는 X_train, X_test, y_train, y_test를 반환합니다.

### make_dataset() 함수 반환값
- 함수가 None을 반환하지 않으며, 실제로 데이터가 잘 반환되고 있습니다.
반환값은 tuple 형식이며, 각 데이터 세트(훈련 세트와 테스트 세트)가 잘 나누어졌습니다.
- 출력된 내용에서 반환값이 tuple임을 확인할 수 있으며,
- 데이터가 [284 rows x 30 columns]와 같은 형태로 나옵니다.

### 에러 메시지 TypeError: cannot unpack non-iterable NoneType object
- 이 오류는 make_dataset() 함수가 None을 반환했다고 예상할 때 발생합니다.
- 그러나 위에서 확인된 결과는 tuple로 정상적으로 데이터가 반환되었습니다

In [23]:
# 타켓 확인
y_train.value_counts()

target
1    190
0     94
Name: count, dtype: int64

In [24]:
# 의사결정나무
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(random_state=0)
model.fit(X_train, y_train)

pred = model.predict(X_test)
accuracy_score(y_test, pred)

0.9263157894736842

### 3. 의사결정나무 하이어파라미터

In [28]:
# 의사결정나무 하이퍼파라미터
from sklearn.tree  import DecisionTreeClassifier

model = DecisionTreeClassifier(
    criterion='entropy',
    max_depth=7,
    min_samples_split=2,
    min_samples_leaf=2,
    random_state=0
)

model.fit(X_train, y_train)
pred = model.predict(X_test)
accuracy_score(y_test, pred)



0.9228070175438596